client/state/query-client-ssr.ts

```diff
@@ -1,13 +1,44 @@
-import { QueryClient, QueryCache } from 'react-query';
+import { QueryClient, QueryCache, dehydrate, QueryOptions, QueryKey } from 'react-query';
 import { MAX_AGE, BASE_STALE_TIME } from 'calypso/state/initial-state';
+import type { Query, QueryState } from 'react-query/types/core/query';
 
-const sharedCache = new QueryCache();
+class QueryClientSSR extends QueryClient {
+	fetchedQueryHashes = new Set< string >();
+}
```

Comment:
One more idea: there's a way to implement this without subclassing `QueryClient` -- storing the hashes in a weakmap.
```js
const fetchedQueryHashes = new WeakMap< QueryClient, Set< string > >;

function addQueryHash( client, queryHash ) {
  let queryHashes = fetchedQueryHashes.get( client );
  if ( ! queryHashes ) {
    queryHashes = new Set();
    fetchedQueryHashes.set( client, queryHashes );
  }
  queryHashes.add( queryHash );
}

function hasQueryHash( client, queryHash ) {
  return fetchedQueryHashes.get( client )?.has( queryHash ) ?? false;
}
```
This code is basically adding a new field to `QueryClient`, without having to extend it. The info for a particular instance is auto-removed when that instance is GC-ed.

We still need to extend `QueryCache` because we're overriding its method.

Reply:
Thanks for the suggestion!
I implemented it and works very well. 
I'm glad we can avoid extending `QueryClient` with this.

With this in mind, I think we can proceed and merge this.